{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f42ab9bf-e63e-48d3-9671-91e21ea445d7",
   "metadata": {},
   "source": [
    "# Chapter 2. Setting Up and Managing a Bioinformatics Project\n",
    "---\n",
    "\n",
    "This is what a typical bioinformatics project direcotry would look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9dedcad2-4dd9-45e5-9977-d717a12f1ce0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 1: cd: ../test-project/: No such file or directory\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m.\u001b[0m\n",
      "├── bioinfx_skills.ipynb\n",
      "├── chapter01.ipynb\n",
      "├── chapter02.ipynb\n",
      "├── foundations_of_git.html\n",
      "└── \u001b[01;34mimages\u001b[0m\n",
      "    ├── \u001b[01;35mawk_comp_logic.png\u001b[0m\n",
      "    ├── \u001b[01;35mawk_funcs.png\u001b[0m\n",
      "    ├── \u001b[01;35mcommit_chain.png\u001b[0m\n",
      "    ├── \u001b[01;35mdiff_output.png\u001b[0m\n",
      "    ├── \u001b[01;35mfindOverlaps.png\u001b[0m\n",
      "    └── \u001b[01;35mgrep_speed.png\u001b[0m\n",
      "\n",
      "1 directory, 10 files\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ../test-project/\n",
    "tree ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20b31ee4-efc8-412e-acfe-454fb024cd62",
   "metadata": {},
   "source": [
    "Above direcotry is a sensible project layout scheme.\n",
    "- *data/* contains all raw and intermediate data.\n",
    "- General project-wide scripts will be in a *scripts/* directory. If scripts contain many files (e.g., multiple Python modules), they should reside in their own\n",
    "subdirectory.\n",
    "- *analysis/* directory contains many smaller analyses—for (e.g. quality of your raw sequences, the aligner output, and the final data that will produce\n",
    "figures and tables for a paper.\n",
    "- Use relative paths to refer to other files becuase using absolute paths leaves your work less portable between collaborators and decreases reproducibility.\n",
    "\n",
    "### Documentation\n",
    "We can keep our documentation in a plain-text *README* file because plain-text files:\n",
    "\n",
    "- Can easily be read, searched, and edited directly from the command line.\n",
    "- Can be put under version control.\n",
    "- Is future-proof format; plain-text files written in the 1960s are still readable today.\n",
    "\n",
    "What exactly do we need to document?\n",
    "\n",
    "- Methods and workflows\n",
    "    - Full command lines that are run through the shell that generate data or intermediate results.\n",
    "    - When using scripts, include command lines options.\n",
    "- Origin of all data in our project directory\n",
    "    - Keep track of where data was downloaded from, who gave it to you, and any other relevant information.\n",
    "- When data was downloaded\n",
    "    - A script that downloads data directly from a database might produce different results if rerun after the external database is updated. Consequently, it’s important to document when data came into your repository.\n",
    "- Data version information\n",
    "    - Many databases have explicit release numbers, version numbers, or names (e.g., TAIR10 version of genome annotation for Arabidopsis thaliana, or Wormbase release WS231 for Caenorhabditis elegans). It’s important to record all version information in your documentation, including minor version numbers.\n",
    "- How data was downloaded\n",
    "    - Was MySQL used to download a set of genes? Or the UCSC Genome Browser? These details can be useful in tracking down issues like when data is different between collaborators.\n",
    "- The versions of the software that is run\n",
    "\n",
    "A good approach is to keep *README* files in each of your project’s main directories. These *README* files don’t necessarily need to be lengthy, but they should at the very least explain what’s in this directory and how it got there. For example, a *data/README* file would contain metadata about your data files in the *data/* directory.\n",
    "\n",
    "#### Shell Expansion Tips\n",
    "*Brace expansion* creates strings by expanding out the comma-separated values inside the braces. Using brace expansion, we can create *test-project* directory with one command:\n",
    "\n",
    "```bash\n",
    "mkdir -p test-project/{analysis, scripts, data/seqs}\n",
    "```\n",
    "\n",
    "Let's create some fake sequencing files and play with them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "af90d58b-ba05-4276-8031-910bce98cd84",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 1: cd: ../test-project/data: No such file or directory\n",
      "touch: cannot touch 'seqs/zmaysA_R1.fastq': No such file or directory\n",
      "touch: cannot touch 'seqs/zmaysA_R2.fastq': No such file or directory\n",
      "touch: cannot touch 'seqs/zmaysB_R1.fastq': No such file or directory\n",
      "touch: cannot touch 'seqs/zmaysB_R2.fastq': No such file or directory\n",
      "touch: cannot touch 'seqs/zmaysC_R1.fastq': No such file or directory\n",
      "touch: cannot touch 'seqs/zmaysC_R2.fastq': No such file or directory\n",
      "ls: cannot access 'seqs': No such file or directory\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'cd ../test-project/data\\ntouch seqs/zmays{A,B,C}_R{1,2}.fastq\\nls -l seqs\\n'' returned non-zero exit status 2.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39mrun_cell_magic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbash\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcd ../test-project/data\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mtouch seqs/zmays\u001b[39m\u001b[38;5;124m{\u001b[39m\u001b[38;5;124mA,B,C}_R\u001b[39m\u001b[38;5;124m{\u001b[39m\u001b[38;5;124m1,2}.fastq\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mls -l seqs\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/Packages/anaconda3/lib/python3.11/site-packages/IPython/core/interactiveshell.py:2517\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2515\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2516\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2517\u001b[0m     result \u001b[38;5;241m=\u001b[39m fn(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m   2519\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2520\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2521\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2522\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m~/Packages/anaconda3/lib/python3.11/site-packages/IPython/core/magics/script.py:154\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    153\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 154\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshebang(line, cell)\n",
      "File \u001b[0;32m~/Packages/anaconda3/lib/python3.11/site-packages/IPython/core/magics/script.py:314\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    309\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    310\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    311\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    312\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    313\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 314\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'cd ../test-project/data\\ntouch seqs/zmays{A,B,C}_R{1,2}.fastq\\nls -l seqs\\n'' returned non-zero exit status 2."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ../test-project/data\n",
    "touch seqs/zmays{A,B,C}_R{1,2}.fastq\n",
    "ls -l seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "62dd55ae-bef3-42cf-b7dc-a86d5cf97193",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 2: cd: ../test-project/data/seqs/: No such file or directory\n",
      "ls: cannot access 'zmays[AB]_R1.fastq': No such file or directory\n",
      "ls: cannot access 'zmays[AB]_R2.fastq': No such file or directory\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b\"# We can use ranges of characters to select certain files\\ncd ../test-project/data/seqs/\\nls -l zmays[AB]_R{1,2}.fastq\\n# Note that character ranges won't work with number. Instead, use brace expansion\\n\"' returned non-zero exit status 2.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39mrun_cell_magic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbash\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m# We can use ranges of characters to select certain files\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mcd ../test-project/data/seqs/\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mls -l zmays[AB]_R\u001b[39m\u001b[38;5;124m{\u001b[39m\u001b[38;5;124m1,2}.fastq\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m# Note that character ranges won\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt work with number. Instead, use brace expansion\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Packages/anaconda3/lib/python3.11/site-packages/IPython/core/interactiveshell.py:2517\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2515\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2516\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2517\u001b[0m     result \u001b[38;5;241m=\u001b[39m fn(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m   2519\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2520\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2521\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2522\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m~/Packages/anaconda3/lib/python3.11/site-packages/IPython/core/magics/script.py:154\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    153\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 154\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshebang(line, cell)\n",
      "File \u001b[0;32m~/Packages/anaconda3/lib/python3.11/site-packages/IPython/core/magics/script.py:314\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    309\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    310\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    311\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    312\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    313\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 314\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b\"# We can use ranges of characters to select certain files\\ncd ../test-project/data/seqs/\\nls -l zmays[AB]_R{1,2}.fastq\\n# Note that character ranges won't work with number. Instead, use brace expansion\\n\"' returned non-zero exit status 2."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# We can use ranges of characters to select certain files\n",
    "cd ../test-project/data/seqs/\n",
    "ls -l zmays[AB]_R{1,2}.fastq\n",
    "# Note that character ranges won't work with number. Instead, use brace expansion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83d06e1e-2210-470a-8e71-329c1c3e880a",
   "metadata": {},
   "source": [
    "In general, it’s best to be as restrictive as possible with wildcards to minimize matches with unintended files.\n",
    "\n",
    "#### Leading Zeros and Sorting\n",
    "Another useful trick is to use leading zeros (e.g., file-0021.txt rather than file-21.txt) when naming files. This is useful because lexicographically sorting files (as ls does) leads to the correct ordering.\n",
    "\n",
    "### Markdown Formatting Basics\n",
    "\n",
    "Here is a basic Markdown document illustrating the format:\n",
    "```\n",
    "# *Zea Mays* SNP Calling\n",
    "\n",
    "We sequenced three lines of *zea mays*, using paired-end\n",
    "sequencing. This sequencing was done by our sequencing core and we\n",
    "received the data on 2013-05-10. Each variety should have **two**\n",
    "sequences files, with suffixes `_R1.fastq` and `_R2.fastq`, indicating\n",
    "which member of the pair it is.\n",
    "\n",
    "## Sequencing Files\n",
    "\n",
    "All raw FASTQ sequences are in `data/seqs/`:\n",
    "\n",
    "    $ find data/seqs -name \"*.fastq\"\n",
    "    data/seqs/zmaysA_R1.fastq\n",
    "    data/seqs/zmaysA_R2.fastq\n",
    "    data/seqs/zmaysB_R1.fastq\n",
    "    data/seqs/zmaysB_R2.fastq\n",
    "    data/seqs/zmaysC_R1.fastq\n",
    "    data/seqs/zmaysC_R2.fastq\n",
    "\n",
    "## Quality Control Steps\n",
    "\n",
    "After the sequencing data was received, our first stage of analysis\n",
    "was to ensure the sequences were high quality. We ran each of the\n",
    "three lines' two paired-end FASTQ files through a quality diagnostic\n",
    "and control pipeline. Our planned pipeline is:\n",
    "\n",
    "1. Create base quality diagnostic graphs.\n",
    "2. Check reads for adapter sequences.\n",
    "3. Trim adapter sequences.\n",
    "4. Trim poor quality bases.\n",
    "\n",
    "Recommended trimming programs:\n",
    "\n",
    "- Trimmomatic\n",
    "- Scythe\n",
    "```\n",
    "\n",
    "Let's create a markdown file named *README.md* and copy above text into it, then convert md file to an HTML file usinf Pandoc. Pandoc can convert between a variety of different markup and output formats:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "161a7233-4198-4649-a413-bbfa64aba244",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 1: cd: ../test-project/: No such file or directory\n",
      "bash: line 40: _R1.fastq: command not found\n",
      "bash: line 40: _R2.fastq: command not found\n",
      "bash: line 40: data/seqs/: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ../test-project/\n",
    "# dump text into notebook.md file\n",
    "echo \"\"\"\n",
    "# *Zea Mays* SNP Calling\n",
    "\n",
    "We sequenced three lines of *zea mays*, using paired-end\n",
    "sequencing. This sequencing was done by our sequencing core and we\n",
    "received the data on 2013-05-10. Each variety should have **two**\n",
    "sequences files, with suffixes `_R1.fastq` and `_R2.fastq`, indicating\n",
    "which member of the pair it is.\n",
    "\n",
    "## Sequencing Files\n",
    "\n",
    "All raw FASTQ sequences are in `data/seqs/`:\n",
    "\n",
    "    $ find data/seqs -name \"*.fastq\"\n",
    "    data/seqs/zmaysA_R1.fastq\n",
    "    data/seqs/zmaysA_R2.fastq\n",
    "    data/seqs/zmaysB_R1.fastq\n",
    "    data/seqs/zmaysB_R2.fastq\n",
    "    data/seqs/zmaysC_R1.fastq\n",
    "    data/seqs/zmaysC_R2.fastq\n",
    "\n",
    "## Quality Control Steps\n",
    "\n",
    "After the sequencing data was received, our first stage of analysis\n",
    "was to ensure the sequences were high quality. We ran each of the\n",
    "three lines' two paired-end FASTQ files through a quality diagnostic\n",
    "and control pipeline. Our planned pipeline is:\n",
    "\n",
    "1. Create base quality diagnostic graphs.\n",
    "2. Check reads for adapter sequences.\n",
    "3. Trim adapter sequences.\n",
    "4. Trim poor quality bases.\n",
    "\n",
    "Recommended trimming programs:\n",
    "\n",
    "- Trimmomatic\n",
    "- Scythe\n",
    "\"\"\" > notebook.md"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f83fea-b611-4b40-bbcb-336db466ad01",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ../test-project/\n",
    "# convert md file to html file using Pandoc\n",
    "pandoc --from markdown --to html notebook.md > output.html\n",
    "pandoc --from html --to pdf output.html --output output.pdf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
